director/sort.go

```diff
@@ -175,42 +169,87 @@ func getClientLatLong(addr netip.Addr) (coord Coordinate, ok bool) {
+//   - distance: sort serverAds by the distance between the geolocation of the servers and the client
+//   - distanceAndLoad: sort serverAds by the distance with gated halving factor (see details in the smart method)
+//     and the server IO load
+//   - random: sort serverAds randomly
+//   - smart:  sort serverAds based on rules discussed here: https://github.com/PelicanPlatform/pelican/discussions/1198
+//
+// Note that if the client has invalid IP address or MaxMind is unable to get the coordinates out of
+// the client IP, any distance-related steps are skipped. If the sort method is "distance", then
+// the serverAds are randomly sorted.
+func sortServerAds(clientAddr netip.Addr, ads []server_structs.ServerAd, availabilityMap map[string]bool) ([]server_structs.ServerAd, error) {
 	// Each entry in weights will map a priority to an index in the original ads slice.
 	// A larger weight is a higher priority.
 	weights := make(SwapMaps, len(ads))
 	sortMethod := param.Director_CacheSortMethod.GetString()
 
-	// If the client addr is not valid, we use random sort
-	if !clientAddr.IsValid() {
-		sortMethod = "random"
+	clientCoord := Coordinate{}
+	clientCoordOk := false
+
+	if clientAddr.IsValid() {
+		clientCoord, clientCoordOk = getClientLatLong(clientAddr)
+	} else {
+		log.Warningf("Unable to sort servers based on client-server distance. Invalid client IP address: %s", clientAddr.String())
 	}
 
 	// For each ad, we apply the configured sort method to determine a priority weight.
 	for idx, ad := range ads {
 		switch sortMethod {
 		case "distance":
-			clientCoord, ok := getClientLatLong(clientAddr)
-			if !ok {
-				// Unable to compute distances for this server; just do random distances.
+			// If we can't get the client coordinates, then simply use random sort
+			if !clientCoordOk {
+				weights[idx] = SwapMap{rand.Float64(), idx}
+				continue
+			}
+			if ad.Latitude == 0 && ad.Longitude == 0 {
+				// If server lat and long are 0, the server geolocation is unknown
 				// Below we sort weights in descending order, so we assign negative value here,
 				// causing them to always be at the end of the sorted list.
 				weights[idx] = SwapMap{0 - rand.Float64(), idx}
 			} else {
-				weights[idx] = SwapMap{distanceWeight(clientCoord, ad),
+				weights[idx] = SwapMap{distanceWeight(clientCoord.Lat, clientCoord.Long, ad.Latitude, ad.Longitude, false),
 					idx}
 			}
 		case "distanceAndLoad":
-			clientCoord, ok := getClientLatLong(clientAddr)
-			if !ok {
-				weights[idx] = SwapMap{0 - rand.Float64(), idx}
-			} else {
-				// Each server ad will have a load value that we can use for sorting
-				weights[idx] = SwapMap{distanceAndLoadWeight(clientCoord, ad),
-					idx}
+			weight := 1.0
+			// Distance weight
+			if clientCoordOk {
+				distance := distanceWeight(clientCoord.Lat, clientCoord.Long, ad.Latitude, ad.Longitude, true)
+				dWeighted := gatedHavlingMultiplier(distance, distanceHalvingThreshold, distanceHalvingFactor)
```

Comment:
There's a typo in the function name -- it should be `gatedHalvingMultiplier` (switched v/l)